lib/Dataflow_types.ml

```diff
@@ -0,0 +1,124 @@
+   * rhs_set: The 'right hand side' set of variables that affect the value or behavior of
+     this node
+   * controlflow: The set of control flow nodes that are immediate parents of this node:
+     * The most recent nested if/then or loop,
+     * or the beginning of the function or block if there are no containing branches,
+     * plus the set of relevant continue/return statements,
+     * plus, for loops, any break statements they contain
+   * loc: The location of the Mir node that this node corresponds to, or a description if
+     there is none
+*)
+type 'rd_info node_info =
+  { rd_sets: 'rd_info
+  ; possible_previous: label Set.Poly.t
+  ; rhs_set: vexpr Set.Poly.t
+  ; controlflow: label Set.Poly.t
+  ; loc: source_loc }
+[@@deriving sexp]
+
+(**
+   A node_info, where the reaching definition information takes the form of an update
+   function that maps from the 'entry' set to the 'exit' set, where the entry set is
+   what's true before executing this node and the exit set is true after.
+*)
+type node_info_update =
+  (reaching_defn Set.Poly.t -> reaching_defn Set.Poly.t) node_info
+
+(**
+   A node_info where the reaching definition information is explicitly written as the
+   entry and exit sets, as after finding the fixed-point solution.
+*)
+type node_info_fixedpoint =
+  (reaching_defn Set.Poly.t * reaching_defn Set.Poly.t) node_info
+[@@deriving sexp]
+
+(**
+   The state that will be maintained throughout the traversal of the Mir
+   * label_ix: The next label that's free to use
+   * node_info_map: The label information that's been built so far
+   * possible_previous: The set of nodes that could have immediately preceded this point
+     under some execution of the program
+   * target_terms: The set of nodes that correspond to terms added to the target variable
+   * continues: A set of the continue nodes that have been encountered since exiting a loop
+   * breaks: A set of the break nodes that have been encountered since exiting a loop
+   * returns: A set of the return nodes that have been encountered
+*)
+type traversal_state =
+  { label_ix: label
+  ; node_info_map: node_info_update Int.Map.t
+  ; possible_previous: label Set.Poly.t
+  ; target_terms: label Set.Poly.t
+  ; continues: label Set.Poly.t
+  ; breaks: label Set.Poly.t
+  ; returns: label Set.Poly.t
+  ; rejects: label Set.Poly.t }
+
+(** The most recently nested control flow (block start, if/then, or loop)
+
+    This isn't included in the traversal_state because it only flows downward through the
+    tree, not across and up like everything else
+*)
+type cf_state = label
+
+(**
+   Everything we need to know to do dependency analysis
+   * node_info_map: Collection of node information
+   * possible_exits: Set of nodes that could be the last to execute under some execution
+   * probabilistic_nodes: Set of nodes corresponding to which can only introduce
+     probabilistic dependencies, such as target terms and reject statements, to be
```

Comment:
@rybern , have you read the language manual section on reject statements? They are a bit crazy in Stan. Generally, they are meant to be exceptions - they cease execution - , but in transformed parameters and model blocks they cause the metropolis proposal to be rejected and do not cease execution altogether. (Just wanted to be sure you were aware of this strange behaviour.)

Reply:
Right, this came from my discussion with Sean. My thinking is that reject statements don't introduce any 'classical' dependencies because they end execution of the step, but that they can introduce probabilistic dependencies. For example, this model block:
```
if (x > 0) {
  if (y > 0) {
    reject();
  }
}
x ~ normal(0,1);
y ~ normal(0,1);
```
I think would be a 2-D normal with a quadrant missing, which means the reject is introducing dependency between `x` and `y`. The way it's handled now is to introduce the variables that are used the control flow surrounding the reject() as dependencies for all other statements. What do you think?

Reply:
Oh, maybe it's just mutual dependency for each variable in the control flow. Hmm

Reply:
Reject statements seem to be introducing a probabilistic dependency between all the control flow dependencies they have, right?

Reply:
I think so yes